language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+use diem_state_view::StateView;
+use diem_transaction_replay::DiemDebugger;
+use diem_types::access_path::AccessPath;
+use diem_types::transaction::TransactionOutput;
+use diem_types::{
+    account_state::AccountState, account_state_blob::AccountStateBlob, transaction::Transaction,
+};
+use diem_vm::{DiemVM, VMExecutor};
+use move_binary_format::CompiledModule;
+use move_cli::sandbox::utils::{
+    mode::{Mode, ModeType},
+    on_disk_state_view::OnDiskStateView,
+};
+use move_core_types::language_storage::ResourceKey;
+
+pub struct OnDiskStateViewReader {
+    view: OnDiskStateView,
+}
+
+struct AccountAddressWithState {
+    address: AccountAddress,
+    state: AccountState,
+}
+
+pub struct PrefetchTransactionReplayerConfig {
+    build_dir: PathBuf,
+    storage_dir: PathBuf,
+    rpc_url: String,
+    module_mode_type: ModeType,
+}
+
+pub struct PrefetchTransactionReplayer {
+    state: OnDiskStateViewReader,
+    debugger: DiemDebugger,
+    rpc_client: BlockingClient,
+}
+
+impl StateView for OnDiskStateViewReader {
+    fn get(&self, access_path: &AccessPath) -> Result<Option<Vec<u8>>> {
+        access_path.get_struct_tag().map_or(Ok(None), |tag| {
+            self.view.get_resource_bytes(access_path.address, tag)
+        })
+    }
+
+    fn is_genesis(&self) -> bool {
+        false
+    }
+}
+
+impl PrefetchTransactionReplayer {
+    pub fn from_config(config: PrefetchTransactionReplayerConfig) -> Result<Self> {
+        let mode = Mode::new(config.module_mode_type);
+        let state = OnDiskStateViewReader {
+            view: mode.prepare_state(&config.build_dir, &config.storage_dir)?,
+        };
+        let debugger = DiemDebugger::json_rpc_with_config(
+            &config.rpc_url,
+            config.build_dir,
+            config.storage_dir,
+        )?;
+
+        let rpc_client = BlockingClient::new(&config.rpc_url);
+
+        Ok(PrefetchTransactionReplayer {
+            state,
+            debugger,
+            rpc_client,
+        })
+    }
+
+    fn get_keys_read_for_tx(
+        &self,
+        tx: &Transaction,
+        modules: &[CompiledModule],
+    ) -> Result<Vec<ResourceKey>> {
+        let signed_tx = match tx {
+            Transaction::UserTransaction(u) => u,
+            _ => bail!("can only process user transactions"),
+        };
+
+        let rw_analysis = ReadWriteSetAnalysis::new(read_write_set::analyze(modules)?);
+        rw_analysis.get_keys_read(signed_tx, &self.state.view)
+    }
+
+    fn get_unique_addresses_from_keys(&self, key_reads: &[ResourceKey]) -> Vec<AccountAddress> {
+        key_reads
+            .iter()
+            .map(|k| k.address)
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect()
+    }
+
+    fn extract_account_state(
+        &self,
+        state_with_proof: &AccountStateWithProofView,
+    ) -> Option<AccountState> {
+        // NOTE: if this actually has a chance to fail, it might be better to change
```

Comment:
I think it's safe to unwrap here, I'm not sure why the blob is an `Option`.

Reply:
Cool, that makes things a bit more readable. As far as I understand, as long as the state is valid, all the operations in there should always succeed so I unwrapped the result too rather than returning it.